Assets/HoloToolkit/Input/Scripts/Utilities/Managers/InputManager.cs

```diff
@@ -61,6 +61,146 @@ public class InputManager : Singleton<InputManager>
+            if (inputSourceTypes.ContainsKey(inputSourceId)) {
+                return inputSourceTypes[inputSourceId];
+            }
+            return MixedRealityInputType.Other;
+        }
+
+        /// <summary>
+        /// Finds the most plausible MixedRealityInputType for an input source and adds it to inputSourceTypes
+        /// TODO: this seems VERY prone to error - find a better way? Bake into IInputSource.TryGetSourceKind?
+        /// </summary>
+        /// <param name="newSource"></param>
+        private void DetermineInputSourceType(InputSourceInfo newSource) {
+            MixedRealityInputType inputType = MixedRealityInputType.Other;                    
+            InteractionSourceInfo sourceKind = InteractionSourceInfo.Other;
+            if (newSource.InputSource.TryGetSourceKind (newSource.SourceId, out sourceKind)) {
+                switch (sourceKind) {
+                    case InteractionSourceInfo.Controller:
+                        // This narrows it down to:
+                        // LeftController
+                        // RightController
+                        // Gamepad
+                        // 
+                        // Use the device's supported input and handedness to determine which it is
+                        if (newSource.InputSource.SupportsInputInfo(newSource.SourceId, SupportedInputInfo.Position)) {
+                            // Get the device's handedness
+                        } else {
+                            // Only gamepad would NOT support position
+                            inputType = MixedRealityInputType.Gamepad;
+                        }
+                        break;
+
+                    case InteractionSourceInfo.Hand:
+                        inputType = MixedRealityInputType.Hand;
+                        break;
+
+                    case InteractionSourceInfo.Voice:
+                        inputType = MixedRealityInputType.Voice;
+                        break;
+
+                    case InteractionSourceInfo.Other:
+                    default:
+                        // This narrows it down to:
+                        // Hand
+                        // Head
+                        //
+                        // Only head supports pointing
+                        // TODO find out if this is true?
+                        if (newSource.InputSource.SupportsInputInfo(newSource.SourceId, SupportedInputInfo.Pointing)) {
+                            inputType = MixedRealityInputType.Head;
+                        } else {
+                            inputType = MixedRealityInputType.Hand;
+                        }
+                        break;
+                }
+            }
+
+            // Add the lookup
+            if (!inputSourceTypes.ContainsKey(newSource.SourceId)) {
+                inputSourceTypes.Add(newSource.SourceId, inputType);
+            } else {
+                inputSourceTypes[newSource.SourceId] = inputType;
```

Comment:
I don't think there's currently anything that checks to make sure new detected sources have the same Id, and potentially different sources can have the same id even though the source kind is different.

Reply:
@keveleigh warned me about about this. Do we have a cheat sheet for expected ID behavior?

Reply:
None, that I'm aware of.

I think we should have the InputManager assign this sourceId as they're detected.

Reply:
I've removed these functions along with SourceType.